apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java

```diff
@@ -814,6 +832,13 @@ public void deactivate(ElasticContext<?> context) {
         if (logger.isDebugEnabled()) {
             logger.debug("Deactivating {} on thread {}", context, Thread.currentThread().getId());
         }
+
+        StackOverflowCounter stackOverflowCounter = activeStackOverflowCounter.get();
+        if (stackOverflowCounter.getOverflowCount() > 0) {
```

Comment:
I wonder whether we should re-size the `ArrayDeque` to it's default size at when de-activating the last span. Otherwise, it may hold a very large array in memory per thread if there's one rogue transaction that gets called once in a while which has a very deep activation stack.

Reply:
I don't think we should overthink this. What I added now is only a safety valve to avoid OOM, not a fix. 
This stack should never grow this much. In the rare cases that they do, it indicates that something is not working as expected for the user. Because we think this is a very rare case, we issue an error-level log and in these rare cases.

Besides, the real effect on memory is when these stacks refer to something, not when they are empty. If you compare a thread with, say 1K-size empty array, to another thread with 500 kept-alive spans through the stack, then this is negligible. And if these spans are kept alive elsewhere as well, then the memory issue is a lot worse and the OOM will occur anyway...